controllers/cloudstackzone_controller.go

```diff
@@ -0,0 +1,120 @@
+	"github.com/aws/cluster-api-provider-cloudstack/pkg/cloud"
+	"github.com/pkg/errors"
+)
+
+// +kubebuilder:rbac:groups=infrastructure.cluster.x-k8s.io,resources=cloudstackzones,verbs=get;list;watch;create;update;patch;delete
+// +kubebuilder:rbac:groups=infrastructure.cluster.x-k8s.io,resources=cloudstackzones/status,verbs=create;get;update;patch
+// +kubebuilder:rbac:groups=infrastructure.cluster.x-k8s.io,resources=cloudstackzones/finalizers,verbs=update
+
+// CloudStackZoneReconciliationRunner is a ReconciliationRunner with extensions specific to CloudStackCluster reconciliation.
+type CloudStackZoneReconciliationRunner struct {
+	csCtrlrUtils.ReconciliationRunner
+	Zones                 *infrav1.CloudStackZoneList
+	ReconciliationSubject *infrav1.CloudStackZone
+	CSUser                cloud.Client
+	IsoNet                *infrav1.CloudStackIsolatedNetwork
+}
+
+// CloudStackZoneReconciler reconciles a CloudStackZone object
+type CloudStackZoneReconciler struct {
+	csCtrlrUtils.ReconcilerBase
+}
+
+// Initialize a new CloudStackZone reconciliation runner with concrete types and initialized member fields.
+func NewCSZoneReconciliationRunner() *CloudStackZoneReconciliationRunner {
+	// Set concrete type and init pointers.
+	r := &CloudStackZoneReconciliationRunner{ReconciliationSubject: &infrav1.CloudStackZone{}}
+	r.Zones = &infrav1.CloudStackZoneList{}
+	r.IsoNet = &infrav1.CloudStackIsolatedNetwork{}
+	// Setup the base r. Initializes pointers and links reconciliation methods.
+	r.ReconciliationRunner = csCtrlrUtils.NewRunner(r, r.ReconciliationSubject)
+	return r
+}
+
+func (reconciler *CloudStackZoneReconciler) Reconcile(ctx context.Context, req ctrl.Request) (res ctrl.Result, retErr error) {
+	return NewCSZoneReconciliationRunner().
+		UsingBaseReconciler(reconciler.ReconcilerBase).
+		ForRequest(req).
+		WithRequestCtx(ctx).
+		RunBaseReconciliationStages()
+}
+
+// SetupWithManager sets up the controller with the Manager.
+func (reconciler *CloudStackZoneReconciler) SetupWithManager(mgr ctrl.Manager) error {
+	return ctrl.NewControllerManagedBy(mgr).
+		For(&infrav1.CloudStackZone{}).
+		Complete(reconciler)
+}
+
+func (r *CloudStackZoneReconciliationRunner) Reconcile() (retRes ctrl.Result, reterr error) {
+	if res, err := r.RequeueIfMissingBaseCRDs(); r.ShouldReturn(res, err) {
+		return res, err
+	}
+	// Prevent premature deletion.
+	controllerutil.AddFinalizer(r.CSCluster, infrav1.ZoneFinalizer)
+
+	r.Log.V(1).Info("Reconciling CloudStackCluster.", "clusterSpec", r.ReconciliationSubject.Spec)
+	// Start by purely data fetching information about the zone and specified network.
+	var res ctrl.Result
+	if err := r.CS.ResolveZone(r.ReconciliationSubject); err != nil {
+		return ctrl.Result{}, errors.Wrap(err, "error encountered when resolving CloudStack zone information")
+	} else if res, err = r.PatchChangesBackToAPI(); r.ShouldReturn(res, err) { // Persist found zone ID.
```

Comment:
No need for this else.  Better to have the primary workflow steps at the function's outer scope.

Reply:
That patch back method isn't even needed with my defer update now. : )